### PR TITLE
Feature: events

### PR DESCRIPTION
**Pull request by Zorin95670**
:information_source:  Don't forget to modify the changelog.md before merging this branch.

### CODE REVIEW

File: guides/migrations/0.14.0_to_X.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  {
+    SelectEvent: null,
+    UpdateEvent: null,
+  },
+  rootId,
+  options,
+);
+
+// After:
+
```
Comment (justinehell) [inline guides/migrations/0.14.0_to_X.md:62]: remove this line

File: guides/migrations/0.14.0_to_X.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  resources,
+  rootId,
+  options,
+);
+```
+
+- Remove methods:
+  - `setEvents`
+  - `emitUpdateEvent`
+- `createLink` method now have `componentId` in parameters
```
Comment (justinehell) [inline guides/migrations/0.14.0_to_X.md:74]: missing `.`

File: changelog.md [hunk 3]
```diff
@@ -5,6 +5,22 @@ All notable changes to this project will be documented in this file.
...
+- Rework on all events.
+- Store of event logs.
+
+### Changed
+
+- Remove old events management in `DefaultDrawer`.
+
+### Fixed
+
+- Prevent dependencies auto-updating in package-lock with `=` suffix.
```
Comment (justinehell) [inline changelog.md:22]: - Prevent auto-updating of dependencies in package-lock with `=` suffix.

File: README.md [hunk 4]
```diff
@@ -96,6 +96,24 @@ This is the default lifecycle of plugin usage in Leto Modelizer.
 
 Create you plugin project from [leto-modelizer-plugin-template](https://github.com/ditrit/leto-modelizer-plugin-template) and follow the readme section of the template project.
 
+### Events
+
+By default plugin send event if you provide a next function in DefaultData like:
```
[the remark this answers is not in the document]
Comment (k-y0u) [reply README.md:101]: sendS

[the remark this answers is not in the document]
Comment (k-y0u) [reply README.md:101]: `next()`

File: README.md [hunk 5]
```diff
@@ -96,6 +96,24 @@ This is the default lifecycle of plugin usage in Leto Modelizer.
...
+new DefaultPlugin({
+  event: {
+    next: () => {},
+  }
+})
+```
+
+If you do not, event are still store in the `DefaultData.eventLogs`.
+
+When you create your plugin you can send event to Leto-modelizer to see progression of your parsing or rendering.
```
Comment (k-y0u) [inline README.md:113]: add sentence for other action.

File: README.md [hunk 6]
```diff
@@ -154,6 +172,8 @@ This is the default directory structure we use for this project:
 leto-modelizer-plugin-core
 ├ docs                 ⇨ Contains all files generated by esdoc
 ├ reports              ⇨ Contains all the report files for sonar
+├ guides               ⇨ Contains all the guides
+│ └  migrations        ⇨ Contains all migrations guides
```
Comment (k-y0u) [inline README.md:176]: `all migration guides`

File: guides/migrations/0.14.0_to_X.md [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+```js
+// Before:
+new DefaultData(props);
+
+// After:
+new Defaultdata(props, { next: () => {} });
+```
+
+- Add new method `emitEvent()`
```
Comment (k-y0u) [inline guides/migrations/0.14.0_to_X.md:41]: missing dot

File: guides/migrations/0.14.0_to_X.md [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+- `DefaultData` contains all event logs in `DefaultData.eventLogs`
```
Comment (k-y0u) [inline guides/migrations/0.14.0_to_X.md:42]: missing dot

**Review by k-y0u — CHANGES_REQUESTED** [review]

File: src/models/DefaultData.js [hunk 9]
```diff
@@ -361,6 +383,70 @@ class DefaultData {
       Math.min(this.components.length - 1, targetIndex + (targetIndex < movedIndex)),
     );
   }
+
+  /**
+   * Get event log by id.
+   *
+   * @param {number} id - Event log id.
+   * @returns {EventLog} Event log or null.
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/DefaultData.js:402]: done

File: src/models/DefaultData.js [hunk 10]
```diff
@@ -361,6 +383,70 @@ class DefaultData {
...
+  /**
+   * Emit event with log.
+   *
+   * @param {object} props - EventLog information.
+   * @returns {number} EventLog id.
+   * @see EventLog
+   */
+  emitEvent(props = {}) {
+    let { id } = props;
+    let event;
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/DefaultData.js:426]: done

File: tests/unit/models/DefaultData.spec.js [hunk 11]
```diff
@@ -45,14 +47,15 @@ describe('Test class: DefaultData', () => {
...
-      });
+      }, {});
 
       expect(pluginData.name).toEqual('name');
       expect(pluginData.defaultFileName).toEqual('default');
       expect(pluginData.version).toEqual('version');
       expect(pluginData.components).toEqual([0]);
       expect(pluginData.parseErrors).toEqual([2]);
       expect(pluginData.definitions).toEqual({ components: [3], links: [4] });
+      expect(pluginData.event).toEqual({ });
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply tests/unit/models/DefaultData.spec.js:58]: done

File: tests/unit/models/DefaultPlugin.spec.js [hunk 12]
```diff
@@ -125,10 +140,36 @@ describe('Test class: DefaultPlugin', () => {
         expect(mockParseConfiguration).toBeCalled();
         expect(mockParse).toBeCalled();
       });
+
+      it('should create 2 success event log', () => {
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply tests/unit/models/DefaultPlugin.spec.js:145]: done

File: src/draw/DefaultDrawer.js [hunk 13]
```diff
@@ -1178,13 +1221,23 @@ class DefaultDrawer {
...
+        },
+      });
     }
   }
 
   /**
    * Create a link between the previously selected source and destination.
    */
-  createLink() {
+  createLink(componentId) {
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/draw/DefaultDrawer.js:1240]: done

File: src/parser/DefaultParser.js [hunk 14]
```diff
@@ -22,8 +22,9 @@ class DefaultParser {
    * Convert the content of files into Components.
    *
    * @param {FileInput[]} [inputs=[]] - Data you want to parse.
+   * @param {string} [parentEventId] - Parent event id.
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/parser/DefaultParser.js:25]: done

File: src/parser/DefaultParser.js [hunk 15]
```diff
@@ -32,15 +33,45 @@ class DefaultParser {
    * Set configuration into Components.
    *
    * @param {FileInput} file - Configuration file of components.
+   * @param {string} [parentEventId] - Parent event id.
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/parser/DefaultParser.js:36]: done

File: src/render/DefaultRender.js [hunk 16]
```diff
@@ -22,16 +22,17 @@ class DefaultRender {
    * Transform all provided components in file inputs.
    *
    * @param {FileInput[]} files - Files managed by the plugin.
+   * @param {string} [parentEventId] - Parent event id.
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/render/DefaultRender.js:25]: done

File: src/render/DefaultRender.js [hunk 17]
```diff
@@ -42,18 +43,31 @@ class DefaultRender {
...
+  // eslint-disable-next-line no-unused-vars
+  renderFiles(parentEventId) {
     return [];
   }
 
   /**
    * Update configuration file content according to components data.
    *
    * @param {FileInput} file - Configuration file of components.
+   * @param {string} [parentEventId] - Parent event id.
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/render/DefaultRender.js:58]: done

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (Zorin95670) [reply README.md:101]: done

File: README.md [hunk 18]
```diff
@@ -96,6 +96,24 @@ This is the default lifecycle of plugin usage in Leto Modelizer.
...
+
+```js
+new DefaultPlugin({
+  event: {
+    next: () => {},
+  }
+})
+```
+
+If you do not, event are still store in the `DefaultData.eventLogs`.
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply README.md:111]: done

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (Zorin95670) [reply README.md:113]: done

File: guides/migrations/0.14.0_to_X.md (hunk 1 above)
Comment (Zorin95670) [reply guides/migrations/0.14.0_to_X.md:62]: done

File: changelog.md (hunk 3 above)
Comment (Zorin95670) [reply changelog.md:22]: done

File: src/models/DefaultData.js [hunk 19]
```diff
@@ -77,6 +80,25 @@ class DefaultData {
...
+     * @type {number}
+     * @private
+     */
+    this.__eventIndex = 0;
+    /**
+     * Event.
+     *
+     * @type {object}
+     */
+    this.event = event;
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/DefaultData.js:95]: done

File: src/models/DefaultData.js [hunk 20]
```diff
@@ -22,6 +23,8 @@ class DefaultData {
    * definitions.
    * @param {ParseError[]} [props.parseErrors=[]] - Parse errors array.
    * @param {string} [props.defaultFileName] - Default file name for new components.
+   * @param {object} [event] - Event.
+   * @param {Function} [event.next] - Function to emit event
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/DefaultData.js:27]: done

File: src/models/DefaultData.js [hunk 21]
```diff
@@ -361,6 +383,70 @@ class DefaultData {
...
+  getEventLogById(id) {
+    return this.eventLogs.findLast((eventLog) => id === eventLog.id);
+  }
+
+  /**
+   * Delete all event logs before the specified datetime.
+   *
+   * @param {number} date - Date time.
+   */
+  deleteAllEventsBefore(date) {
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/DefaultData.js:413]: done

File: src/models/DefaultData.js [hunk 22]
```diff
@@ -361,6 +383,70 @@ class DefaultData {
...
+   * @returns {EventLog} Event log or null.
+   */
+  getEventLogById(id) {
+    return this.eventLogs.findLast((eventLog) => id === eventLog.id);
+  }
+
+  /**
+   * Delete all event logs before the specified datetime.
+   *
+   * @param {number} date - Date time.
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/DefaultData.js:411]: done

File: src/models/DefaultData.js [hunk 23]
```diff
@@ -361,6 +383,70 @@ class DefaultData {
...
+    let event;
+
+    if (!id) {
+      this.__eventIndex += 1;
+      id = this.__eventIndex;
+    } else {
+      event = this.getEventLogById(id);
+    }
+
+    if (!event) {
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/DefaultData.js:435]: done

File: src/models/DefaultData.js [hunk 24]
```diff
@@ -361,6 +383,70 @@ class DefaultData {
...
+      event = new EventLog({ ...props, id });
+      this.eventLogs.push(event);
+    } else {
+      Object.keys(props).forEach((key) => {
+        event[key] = props[key];
+      });
+    }
+
+    if (!event.startDate) {
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/DefaultData.js:444]: done

File: src/models/DefaultData.js [hunk 25]
```diff
@@ -361,6 +383,70 @@ class DefaultData {
...
+      event.startDate = Date.now();
+    }
+
+    if (['success', 'warning', 'error'].includes(event.status)) {
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/DefaultData.js:448]: done

File: src/models/DefaultPlugin.js [hunk 26]
```diff
@@ -12,13 +12,16 @@ class DefaultPlugin {
    * Default constructor.
    *
    * @param {object} [props={}] - Object that contains all properties to set.
+   * @param {object} [event] - Event.
+   * @param {Function} [event.next] - Function to emit event
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/DefaultPlugin.js:16]: done

File: src/models/DefaultData.js [hunk 27]
```diff
@@ -77,6 +80,25 @@ class DefaultData {
      * @type {string}
      */
     this.defaultFileName = props.defaultFileName || null;
+    /**
+     * Start index of event.
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/DefaultData.js:84]: done

File: src/models/EventLog.js [hunk 28]
```diff
@@ -0,0 +1,115 @@
+/**
+ * Event log.
+ */
+class EventLog {
+  /**
+   * Default constructor.
+   *
+   * @param {object} [props={}] - Object that contains all properties to set.
+   * @param {string} [props.id] - EventLog id.
+   * @param {string} [props.parent] - Parent id, to attach this log to another, like a
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/EventLog.js:10]: done

File: src/models/EventLog.js [hunk 29]
```diff
@@ -0,0 +1,115 @@
...
+   * sub-process.
+   * @param {string[]} [props.files] - Impacted files path.
+   * @param {string[]} [props.components] - Impacted components id.
+   * @param {string[]} [props.links] - Impacted links id.
+   * @param {string} [props.type] - Type of log, can be 'Plugin', 'Render', 'Metadata',
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/EventLog.js:15]: done

File: src/models/EventLog.js [hunk 30]
```diff
@@ -0,0 +1,115 @@
...
+   * 'Drawer', 'Parser' or 'Data'.
+   * @param {string} [props.action] - Action of log, can be 'init', 'read', 'write', 'add',
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/EventLog.js:17]: done

File: src/models/EventLog.js [hunk 31]
```diff
@@ -0,0 +1,115 @@
...
+   * 'update', 'move', 'select' or 'delete'.
+   * @param {string} [props.status] - Status of log, can be 'running', 'success', 'warning'
+   * or 'error'.
+   * @param {string | number | Array | object} [props.data] - Extra data of log.
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/EventLog.js:21]: any it's not js :(

File: src/models/EventLog.js [hunk 32]
```diff
@@ -0,0 +1,115 @@
...
+     */
+    this.id = id || null;
+    /**
+     * Parent id, to attach this log to another, like a sub-process.
+     *
+     * @type {string}
+     */
+    this.parent = parent || null;
+    /**
+     * Date time of event creation.
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/EventLog.js:59]: done

File: src/models/EventLog.js [hunk 33]
```diff
@@ -0,0 +1,115 @@
...
+     *
+     * @type {number}
+     */
+    this.startDate = null;
+    /**
+     * Date time of event end.
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/EventLog.js:65]: done

File: src/models/EventLog.js [hunk 34]
```diff
@@ -0,0 +1,115 @@
...
+     *
+     * @type {number}
+     */
+    this.endDate = null;
+    /**
+     * Impacted files path.
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/EventLog.js:71]: done

File: src/models/EventLog.js [hunk 35]
```diff
@@ -0,0 +1,115 @@
...
+     *
+     * @type {string[]}
+     */
+    this.files = files || null;
+    /**
+     * Impacted components id.
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/EventLog.js:77]: done

File: src/models/EventLog.js [hunk 36]
```diff
@@ -0,0 +1,115 @@
...
+     */
+    this.links = links || null;
+    /**
+     * Type of log, can be 'Plugin', 'Render', 'Metadata', 'Drawer', 'Parser' or 'Data'.
+     *
+     * @type {string}
+     */
+    this.type = type || null;
+    /**
+     * Action of log, can be 'init', 'read', 'write', 'add', 'update', 'move', 'select' or 'delete'.
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/EventLog.js:95]: done

File: src/models/EventLog.js [hunk 37]
```diff
@@ -0,0 +1,115 @@
...
+     *
+     * @type {string}
+     */
+    this.action = action || null;
+    /**
+     * Status of log, can be 'running', 'success', 'warning' or 'error'.
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/EventLog.js:101]: done

File: src/models/EventLog.js [hunk 38]
```diff
@@ -0,0 +1,115 @@
...
+     *
+     * @type {string|null}
+     */
+    this.status = status || null;
+    /**
+     * Extra data of log.
+     *
+     * @type {string | number | Array | object}
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/EventLog.js:109]: done

File: src/models/EventLog.js [hunk 39]
```diff
@@ -0,0 +1,115 @@
...
+    /**
+     * Action of log, can be 'init', 'read', 'write', 'add', 'update', 'move', 'select' or 'delete'.
+     *
+     * @type {string}
+     */
+    this.action = action || null;
+    /**
+     * Status of log, can be 'running', 'success', 'warning' or 'error'.
+     *
+     * @type {string|null}
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/EventLog.js:103]: done

File: tests/unit/models/DefaultPlugin.spec.js [hunk 40]
```diff
@@ -125,10 +140,36 @@ describe('Test class: DefaultPlugin', () => {
...
+        const plugin = new DefaultPlugin({
+          pluginData,
+          pluginMetadata: {
+            parse: mockParse,
+          },
+        });
+
+        plugin.parse({ path: 'a' }, [{ path: 'b' }, { path: 'c' }]);
+
+        expect(plugin.data.eventLogs.length).toEqual(2);
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply tests/unit/models/DefaultPlugin.spec.js:157]: done

File: tests/unit/models/DefaultPlugin.spec.js [hunk 41]
```diff
@@ -144,6 +185,28 @@ describe('Test class: DefaultPlugin', () => {
...
+          pluginData,
+          pluginRenderer: {
+            render: mockRender,
+            renderConfiguration: mockRenderConfiguration,
+          },
+        });
+
+        plugin.render({ path: 'a' }, [{ path: 'b' }, { path: 'c' }]);
+
+        expect(plugin.data.eventLogs.length).toEqual(1);
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply tests/unit/models/DefaultPlugin.spec.js:204]: done

File: README.md (hunk 5 above)
Comment (Zorin95670) [reply README.md:113]: please give me the sentence !

File: README.md (hunk 6 above)
Comment (Zorin95670) [reply README.md:176]: done

File: guides/migrations/0.14.0_to_X.md (hunk 7 above)
Comment (Zorin95670) [reply guides/migrations/0.14.0_to_X.md:41]: done

File: guides/migrations/0.14.0_to_X.md (hunk 8 above)
Comment (Zorin95670) [reply guides/migrations/0.14.0_to_X.md:42]: done

File: src/models/DefaultData.js [hunk 42]
```diff
@@ -361,6 +394,66 @@ class DefaultData {
...
+   * @param {object} props - EventLog information.
+   * @returns {number} EventLog id.
+   * @see EventLog
+   */
+  emitEvent(props = {}) {
+    let { id } = props;
+    let eventLog;
+
+    if (!id) {
+      this.__eventIndex += 1;
```
Comment (k-y0u) [inline src/models/DefaultData.js:429]: useless

File: src/models/EventLog.js (hunk 30 above)
[the remark this answers is not in the document]
Comment (k-y0u) [reply src/models/EventLog.js]: missing dot

File: src/models/EventLog.js (hunk 34 above)
[the remark this answers is not in the document]
Comment (k-y0u) [reply src/models/EventLog.js]: not in props description

File: src/models/EventLog.js (hunk 35 above)
[the remark this answers is not in the document]
Comment (k-y0u) [reply src/models/EventLog.js]: not in props description

File: src/models/EventLog.js [hunk 43]
```diff
@@ -0,0 +1,115 @@
...
+     */
+    this.files = files || null;
+    /**
+     * Impacted components id.
+     *
+     * @type {string[]}
+     */
+    this.components = components || null;
+    /**
+     * Impacted links id.
```
[the remark this answers is not in the document]
Comment (k-y0u) [reply src/models/EventLog.js]: not in props description

**Review by k-y0u — CHANGES_REQUESTED** [review]

File: src/models/EventLog.js (hunk 30 above)
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/EventLog.js]: done

File: src/models/EventLog.js (hunk 34 above)
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/EventLog.js]: done

File: src/models/EventLog.js (hunk 35 above)
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/EventLog.js]: done

File: src/models/EventLog.js (hunk 43 above)
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/EventLog.js]: done

File: src/models/DefaultData.js (hunk 42 above)
Comment (Zorin95670) [reply src/models/DefaultData.js:429]: NOPE

**Review by k-y0u — DISMISSED** [review]

**Review by justinehell — DISMISSED** [review]

**Review by k-y0u — APPROVED** [review]

**Review by justinehell — APPROVED** [review]

**Pull request merged**